frontend/app/components/wp-relations/wp-relations-create/wp-relations-autocomplete/wp-relations-autocomplete.directive.ts

```diff
@@ -0,0 +1,129 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+//
+// See doc/COPYRIGHT.rdoc for more details.
+//++
+
+import {wpDirectivesModule} from '../../../../angular-modules';
+import {WorkPackageRelationsController} from "../../wp-relations.directive";
+import {WorkPackageRelationsHierarchyController} from "../../wp-relations-hierarchy/wp-relations-hierarchy.directive";
+import {WorkPackageResourceInterface} from "../../../api/api-v3/hal-resources/work-package-resource.service";
+
+function wpRelationsAutocompleteDirective($q, PathHelper, $http) {
+  return {
+    restrict: 'E',
+    templateUrl: '/components/wp-relations/wp-relations-create/wp-relations-autocomplete/wp-relations-autocomplete.template.html',
+    require: ['^wpRelations', '?^wpRelationsHierarchy'],
+    scope: {
+      selectedWpId: '=',
+      selectedRelationType: '=',
+      workPackage: '='
+    },
+    link: function (scope, element, attrs, controllers ) {
+      scope.relatedWps = [];
+      getRelatedWorkPackages();
+
+      scope.onSelect = function(wpId){
+        scope.selectedWpId = wpId;
+      };
+
+      scope.autocompleteWorkPackages = (term) => {
+        if (!term) {
+          return;
+        }
+
+        findRelatableWorkPackages(term).then((workPackages:Array<WorkPackageResourceInterface>) => {
+          // reject already related work packages, self, children and parent
+          // to prevent invalid relations
+          scope.options = _.reject(workPackages, (wp) => {
+            return scope.relatedWps.indexOf(parseInt((wp.id as string))) > -1;
+          });
+        });
+      };
+
+      function findRelatableWorkPackages(search:string) {
+        const deferred = $q.defer();
+        var params;
+
+        scope.workPackage.project.$load().then(() => {
+          params = {
+            q: search,
+            scope: 'relatable',
+            escape: false,
+            id: scope.workPackage.id,
+            project_id: scope.workPackage.project.id
+          };
+
+          $http({
+            method: 'GET',
+            url: URI(PathHelper.workPackageJsonAutoCompletePath()).search(params).toString()
+          })
+            .then((response:any) => {
+              // THE JSON AUTOCOMPLETE SHOULD BE EXTENDED TO CONTAIN A REFERENCE TO THE
+              // ACTUAL WP-TYPE SINCE MILESTONES MAY NOT BE A PARENT ELEMENT AND THEREFORE THEY
+              // MUST BE REJECTED
+              deferred.resolve(response.data);
+            })
+            .catch(deferred.reject);
+        })
+          .catch(deferred.reject);
+
+        return deferred.promise;
```

Comment:
wouldn't that break the whole sense of an own method for finding relatable wps?

I mean if I get you right, the method would just look like 
```
function findRelatableWorkPackages() {
       return scope.workPackage.project.$load();
}
```

and then I would have all the rest of the nested promises within the `then` block of the `findRelatableWorkPackages()` call.

Since this is a part I only copied from dev I don't know if the autocomplete request could be simplified even further..